lib/class-wp-rest-annotations-controller.php

```diff
@@ -0,0 +1,1097 @@
+				'status' => rest_authorization_required_code(),
+			) );
+		}
+
+		foreach ( array(
+			'id',
+		) as $param ) {
+			if ( isset( $request[ $param ] ) && ( ! isset( $default_params[ $param ] ) || $request[ $param ] !== $default_params[ $param ] ) ) {
+				return new WP_Error( 'rest_readonly_annotation_param_' . $param, __( 'Read-only param.', 'gutenberg' ), array(
+					'status' => 400,
+				) );
+			}
+		}
+
+		// @codeCoverageIgnoreStart
+		// Not possible to test at this time.
+		// Strengthens security in custom post types and on sites with permission mods.
+		if ( ! current_user_can( 'edit_annotations', $type ) ) {
+			foreach ( array(
+				'date',
+				'date_gmt',
+				'meta',
+			) as $param ) {
+				if ( isset( $request[ $param ] ) && ( ! isset( $default_params[ $param ] ) || $request[ $param ] !== $default_params[ $param ] ) ) {
+					return new WP_Error( 'rest_forbidden_annotation_param_' . $param, __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+						'status' => rest_authorization_required_code(),
+					) );
+				}
+			}
+		} // @codeCoverageIgnoreEnd
+
+		if ( ! current_user_can( 'edit_others_annotations', $type ) ) {
+			foreach ( array(
+				'author',
+				'author_ip',
+				'author_user_agent',
+			) as $param ) {
+				if ( isset( $request[ $param ] ) && ( ! isset( $default_params[ $param ] ) || $request[ $param ] !== $default_params[ $param ] ) ) {
+					return new WP_Error( 'rest_forbidden_annotation_param_' . $param, __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+						'status' => rest_authorization_required_code(),
+					) );
+				}
+			}
+
+			$allow_statuses = array( 'approved', 'approve', '1' );
+			$allow_statuses = array_merge( $allow_statuses, WP_Annotation_Utils::$custom_statuses );
+
+			if ( ! in_array( $status, $allow_statuses, true ) ) {
+				return new WP_Error( 'rest_forbidden_annotation_param_status', __( 'Sorry, you are not allowed as this user.', 'gutenberg' ), array(
+					'status' => rest_authorization_required_code(),
+				) );
+			}
+		}
+
+		return true;
+	}
+
+	/**
+	 * Checks if request has access to read items.
+	 *
+	 * @since [version]
+	 *
+	 * @param  WP_REST_Request $request Full REST API request details.
+	 *
+	 * @return bool|WP_Error            True if request has access to read,
+	 *                                  {@see WP_Error} otherwise.
+	 *
+	 * @see WP_REST_Comments_Controller::get_items_permissions_check()
+	 */
+	public function get_items_permissions_check( $request ) {
+		$post_ids       = (array) $request['post'];
+		$types          = (array) $request['type'];
+		$statuses       = (array) $request['status'];
+		$default_params = $request->get_default_params();
+
+		if ( ! $post_ids ) {
+			return new WP_Error( 'rest_missing_annotation_posts', __( 'Missing post(s).', 'gutenberg' ), array(
+				'status' => 400,
```

Comment:
I agree with `400` error codes being out of place here.

The REST API is rather interesting in its approach to permissions. To properly check permissions one must take the current `WP_REST_Request` into consideration, which is a variable that's passed into these permission checks.

The only problem with this is that [`rest_send_allow_header()`](https://developer.wordpress.org/reference/functions/rest_send_allow_header/) filters every API response, and it calls each of the permission checks with the current `WP_REST_Request`, which may or may not contain the type of request parameters expected; e.g., it may call `get_items_permissions_check( $request )` with request parameters originally intended to create a new annotation.

That makes it more difficult to write a permission check that considers the current request, because at times the current request contains, or doesn't contain, all the information necessary to check permissions properly.

There might be something I can do to improve this though.